dali/kernels/imgproc/flip_gpu.cuh

```diff
@@ -27,40 +27,55 @@ namespace kernels {
+                           size_t seq_length, size_t depth, size_t height, size_t width,
+                           size_t channels,
                            bool flip_z, bool flip_y, bool flip_x) {
   size_t xc = blockIdx.x * blockDim.x + threadIdx.x;
   size_t y = blockIdx.y * blockDim.y + threadIdx.y;
-  size_t z = blockIdx.z * blockDim.z + threadIdx.z;
-  if (xc >= width * channels || y >= height || z >= layers) {
+  size_t fz = blockIdx.z * blockDim.z + threadIdx.z;
+  if (xc >= width * channels || y >= height || fz >= depth * seq_length) {
     return;
   }
-  size_t channel = C ? xc % C : xc % channels;
-  size_t x = C ? xc / C : xc / channels;
-  size_t in_x = flip_x ? width - 1 - x : x;
-  size_t in_y = flip_y ? height - 1 - y : y;
-  size_t in_z = flip_z ? layers - 1 - z : z;
-  size_t input_idx = channel + (C ? C : channels) * (in_x + width * (in_y + height * in_z));
-  size_t output_idx = channel + (C ? C : channels) * (x + width * (y + height * z));
+  const size_t channel = C ? xc % C : xc % channels;
+  const size_t x = C ? xc / C : xc / channels;
+  const size_t z = (!Single) ? fz % depth : fz;
+  const size_t f = (!Single) ? fz / depth: 0;
+  const size_t in_x = flip_x ? width - 1 - x : x;
+  const size_t in_y = flip_y ? height - 1 - y : y;
+  const size_t in_z = flip_z ? depth - 1 - z : z;
+  const size_t input_idx =
+      channel + (C ? C : channels) * (in_x + width * (in_y + height * (in_z + depth * f)));
+  const size_t output_idx =
+      channel + (C ? C : channels) * (x + width * (y + height * (z + depth * f)));
   output[output_idx] = input[input_idx];
 }
 
 template <typename T>
 void FlipImpl(T *__restrict__ output, const T *__restrict__ input,
-              size_t layers, size_t height, size_t width, size_t channels,
+              size_t seq_length, size_t depth, size_t height, size_t width, size_t channels,
               bool flip_z, bool flip_y, bool flip_x, cudaStream_t stream) {
-  auto layer_width = width * channels;
-  unsigned int block_x = layer_width < 32 ? layer_width : 32;
+  auto plane_width = width * channels;
+  unsigned int block_x = plane_width < 32 ? plane_width : 32;
   unsigned int block_y = height < 32 ? height : 32;
   dim3 block(block_x, block_y, 1);
-  dim3 grid((layer_width + block_x - 1) / block_x, (height + block_y - 1) / block_y, layers);
-  VALUE_SWITCH(channels, c_channels, (1, 2, 3, 4, 5, 6, 7, 8), (
-      detail::gpu::FlipKernel<c_channels><<<grid, block, 0, stream>>>
-        (output, input, layers, height, width, channels, flip_z, flip_y, flip_x);), (
-      detail::gpu::FlipKernel<0><<<grid, block, 0, stream>>>
-        (output, input, layers, height, width, channels, flip_z, flip_y, flip_x);));
+  dim3 grid((plane_width + block_x - 1) / block_x,
+            (height + block_y - 1) / block_y,
+            depth * seq_length);
+  if (seq_length == 1) {
+    VALUE_SWITCH(channels, c_channels, (1, 2, 3, 4, 5, 6, 7, 8), (
```

Comment:
Do we need all this values in this switch? Maybe some most frequently used should be listed?

Reply:
I think these are the "most frequently used" - we could benchmark how they affect the build time, but I don't think it's a lot. What _could_ be done is to coerce T into integer type, so we only have 4 types (1, 2, 4, 8 bytes) instead of 12 ((u)int8-64, bool, half, float, double)

Reply:
Just asking. Fine for me.